ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+written. This means a separate data structure doesn't need to be maintained
+elsewhere for fast writes and lookups: each node written has enough information
+to look up any other key quickly and otherwise navigate the database.
+
+Each node stores the following basic information:
+
+- `key`: the key that is being created or modified. e.g. `/home/sww/dev.md`
+- `value`: the value stored at that key.
+- `seq`: the sequence number of this entry in the owner's hypercore. 0 is the
+  first, 1 the second, and so forth.
+- `feed`: the ID of the hypercore writer that wrote this
+- `path`: a 2-bit hash sequence of the key's components
+- `trie`: a navigation structure used with `path` to find a desired key
+- `clock`: vector clock to determine node insertion causality
+- `feeds`: an array of { feedKey, seq } for decoding a `clock`
+
+### Vector clock
+
+Each node stores a [vector clock](https://en.wikipedia.org/wiki/Vector_clock) of
+the last known sequence number from each feed it knows about. This is what forms
+the DAG structure.
+
+For example, Alice's vector clock on her seq=3 entry above would be `[3, 0]`
+since she knows of her latest entry (seq=3) and Bob's (seq=0).
+
+*TODO: is this correct? ^^^*
+
+### Prefix trie
+
+Given a HyperDB with hundreds of entries, how can a key like `/a/b/c` be looked
+up quickly?
+
+Each node stores a *prefix [trie](https://en.wikipedia.org/wiki/Trie)* that
+assists with finding the shortest path to the desired key.
+
+When a node is written, its *prefix hash* is computed. This done by first
+splitting the key into its components (`a`, `b`, and `c` for `/a/b/c`), and then
+hashing each component into a 32-character hash, where one character is a 2-bit
+value (0, 1, 2, or 3). The `prefix` hash for `/a/b/c` is
+
+```js
+node.path = [
+1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+4 ]
+```
+
+Each component is divided by a newline. `4` is a special value indicating the
+end of the prefix.
+
+#### Example
+
+Consider a fresh HyperDB. We write `/a/b = 24` and get back this node:
+
+```js
+{ key: '/a/b',
+  value: '24',
+  clock: [ 0 ],
+  trie: [],
+  feeds: [ [Object] ],
+  feedSeq: 0,
+  feed: 0,
+  seq: 0,
+  path: 
+   [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+     0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+     4 ] } 
+```
+
+If you compare this path to the one for `/a/b/c` above, you'll see that the
+first 64 2-bit characters match. This is because `/a/b` is a prefix of `/a/b/c`.
+
+Since this is the first entry, `seq` is 0. Since this is the only known feed,
+`feed` is also 0. `feeds` is an array of entries of the form `{ key: Buffer,
```

Comment:
does this mean that no hypercore writer has an id of 0?

Reply:
The local writer of a hyperdb always has feed id 0. The IDs are subjective (differ from each peer's perspective). When replication happens between two peers each peer exchanges feeds in terms of the actual feed key, and not the ID. The numeric ID just an optimization to keep the incremental indexes small and fast.